lib/location/location_service_utils.c

```diff
@@ -0,0 +1,117 @@
+ *	GlobalSign Root CA - R3
+ *	CN=GlobalSign
+ *	O=GlobalSign
+ *	OU=GlobalSign Root CA - R3
+ *	Serial number=04:00:00:00:00:01:21:58:53:08:A2
+ *	Valid from=18 March 2009
+ *	Valid to=18 March 2029
+ *	Download url=https://secure.globalsign.com/cacert/root-r3.crt
+ */
+static const char tls_certificate_here[] =
+	"-----BEGIN CERTIFICATE-----\n"
+	"MIIDXzCCAkegAwIBAgILBAAAAAABIVhTCKIwDQYJKoZIhvcNAQELBQAwTDEgMB4G\n"
+	"A1UECxMXR2xvYmFsU2lnbiBSb290IENBIC0gUjMxEzARBgNVBAoTCkdsb2JhbFNp\n"
+	"Z24xEzARBgNVBAMTCkdsb2JhbFNpZ24wHhcNMDkwMzE4MTAwMDAwWhcNMjkwMzE4\n"
+	"MTAwMDAwWjBMMSAwHgYDVQQLExdHbG9iYWxTaWduIFJvb3QgQ0EgLSBSMzETMBEG\n"
+	"A1UEChMKR2xvYmFsU2lnbjETMBEGA1UEAxMKR2xvYmFsU2lnbjCCASIwDQYJKoZI\n"
+	"hvcNAQEBBQADggEPADCCAQoCggEBAMwldpB5BngiFvXAg7aEyiie/QV2EcWtiHL8\n"
+	"RgJDx7KKnQRfJMsuS+FggkbhUqsMgUdwbN1k0ev1LKMPgj0MK66X17YUhhB5uzsT\n"
+	"gHeMCOFJ0mpiLx9e+pZo34knlTifBtc+ycsmWQ1z3rDI6SYOgxXG71uL0gRgykmm\n"
+	"KPZpO/bLyCiR5Z2KYVc3rHQU3HTgOu5yLy6c+9C7v/U9AOEGM+iCK65TpjoWc4zd\n"
+	"QQ4gOsC0p6Hpsk+QLjJg6VfLuQSSaGjlOCZgdbKfd/+RFO+uIEn8rUAVSNECMWEZ\n"
+	"XriX7613t2Saer9fwRPvm2L7DWzgVGkWqQPabumDk3F2xmmFghcCAwEAAaNCMEAw\n"
+	"DgYDVR0PAQH/BAQDAgEGMA8GA1UdEwEB/wQFMAMBAf8wHQYDVR0OBBYEFI/wS3+o\n"
+	"LkUkrk1Q+mOai97i3Ru8MA0GCSqGSIb3DQEBCwUAA4IBAQBLQNvAUKr+yAzv95ZU\n"
+	"RUm7lgAJQayzE4aGKAczymvmdLm6AC2upArT9fHxD4q/c2dKg8dEe3jgr25sbwMp\n"
+	"jjM5RcOO5LlXbKr8EpbsU8Yt5CRsuZRj+9xTaGdWPoO4zzUhw8lo/s7awlOqzJCK\n"
+	"6fBdRoyV3XpYKBovHd7NADdBj+1EbddTKJd+82cEHhXXipa0095MJ6RMG3NzdvQX\n"
+	"mcIfeg7jLQitChws/zyrVQ4PkX4268NXSb7hLi18YIvDQVETI53O9zJrlAGomecs\n"
+	"Mx86OyXShkDOOyyGeMlhLxS67ttVb9+E7gUJTb0o2HLO02JQZR7rkpeDMdmztcpH\n"
+	"WD9f\n"
+	"-----END CERTIFICATE-----\n";
+
+const char *location_service_utils_get_ca_certificate_here(void)
+{
+	return tls_certificate_here;
+}
+
+const char *location_service_utils_get_ca_certificate(enum location_service service)
+{
+#if defined(CONFIG_LOCATION_SERVICE_HERE)
+	if (service == LOCATION_SERVICE_HERE) {
+		return location_service_utils_get_ca_certificate_here();
+	}
+#endif
+	return NULL;
+}
+
+static int location_service_utils_provision_certificate(
+	int sec_tag,
+	const char *certificate)
+{
+	int err;
+	bool exists;
+
+	if (certificate == NULL) {
+		LOG_ERR("No certificate was provided by the location service");
+		return -EFAULT;
+	}
+
+	err = modem_key_mgmt_exists(sec_tag, MODEM_KEY_MGMT_CRED_TYPE_CA_CHAIN, &exists);
+	if (err) {
+		LOG_ERR("Failed to check for certificates err %d", err);
+		return err;
+	}
+
+	if (exists) {
+		LOG_DBG("A certificate is already provisioned to sec tag %d", sec_tag);
+		return 0;
+	}
+
+	LOG_INF("Provisioning certificate");
+
+	/*  Provision certificate to the modem */
+	err = modem_key_mgmt_write(sec_tag,
+				   MODEM_KEY_MGMT_CRED_TYPE_CA_CHAIN,
+				   certificate,
+				   strlen(certificate));
+	if (err) {
+		LOG_ERR("Failed to provision certificate, err %d", err);
+		return err;
+	}
+
+	return 0;
+}
+
+int location_service_utils_provision_ca_certificates(void)
+{
+	int ret = -ENOTSUP;
```

Comment:
This function is currently called only when HERE is enabled. If it would be called when HERE is not enabled, it would return `-ENOTSUP`, although technically no error occurred. Because the function is generic as such (based on the function name), I think it would be better to make it return `0` in that case instead. You could even consider removing flagging from where the function is being called, although I'm not sure whether the related error handling code would get optimized out when only nRF Cloud is enabled.

Reply:
So two things:

1. `-ENOTSUP`: I kind of agree to change this to zero but `-ENOTSUP` is good that it raises an error trace in the caller side so developer can try to react on the misconfiguration etc. which has caused that return code. Maybe we could handle it in a cleaner way within `location_service_utils.c`.
2. Remove HERE flagging from `method_cellular.c`: We could do that but then we'd need to flag basically entire `location_service_utils.c` with HERE flagging except `location_service_utils_provision_ca_certificates()` function call. Flagging `location_service_utils_get_ca_certificate()` looks particularly clumsy but you need to do it because otherwise you get compiler warning. Also, flash consumption increases by a small amount (269 bytes).

Reply:
1. But is it possible to configure it so, that you get `-ENOTSUP` currently? If both the function call and the function implementation are flagged with `CONFIG_LOCATION_SERVICE_HERE`, I don't think that's possible.
2. The flagging doesn't look that complicated to me, you have flagging inside function `location_service_utils_provision_ca_certificates()` and everything else in the file can be flagged with a single `#if`. The flash consumption increase probably comes mostly from the error and warning traces.

I guess my main point is that now there's flagging both where location_service_utils_provision_ca_certificates() is being called and in the function itself, along with error handling for a case which to me looks like it can never happen, so it seems a bit excessive. :)

Reply:
As discussed offline, location_service_utils.c will be compiled only when `CONFIG_LOCATION_SERVICE_HERE` is enabled. All `CONFIG_LOCATION_SERVICE_HERE` flagging was removed from that file.